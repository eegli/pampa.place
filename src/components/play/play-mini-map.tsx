import CloseIcon from '@mui/icons-material/Close';
import OpenInNewIcon from '@mui/icons-material/OpenInNew';
import {Box, Button, IconButton} from '@mui/material';
import {forwardRef, ReactNode} from 'react';

type MiniMapProps = {
  enlarged: boolean;
  isPositionSelected: boolean;
  successCallback: () => void;
  toggle: () => void;
  children: ReactNode;
};

export const MiniMap = forwardRef<HTMLDivElement, MiniMapProps>(
  function MiniMapWithRef(props, ref) {
    const {enlarged, isPositionSelected, successCallback, toggle, children} =
      props;

    return (
      <Box
        ref={ref}
        id="mini-map"
        role="region"
        position="absolute"
        bottom={50}
        right={30}
        maxHeight="70%"
        maxWidth="85%"
        minHeight="20%"
        minWidth="30%"
        height={enlarged ? 700 : 150}
        width={enlarged ? 700 : 200}
        sx={{
          transition: '0.2s ease',
        }}
        // Display on top of speed dial
        zIndex={1000}
      >
        <IconButton
          aria-label="mini-map open button"
          onClick={toggle}
          size="large"
          sx={{
            position: 'absolute',
            zIndex: 1000,
            display: enlarged ? 'none' : 'block',
            height: '100%',
            width: '100%',
            background:
              'linear-gradient(rgba(0, 0, 0, 0.6), rgba(0, 0, 0, 0.6))',
            borderRadius: 0,
          }}
        >
          <OpenInNewIcon fontSize="large" />
        </IconButton>
        {children}

        {enlarged ? (
          <>
            <IconButton
              aria-label="mini-map close button"
              onClick={toggle}
              size="large"
              sx={{
                position: 'absolute',
                top: '2%',
                right: '2%',
                background:
                  'linear-gradient(rgba(0, 0, 0, 0.4), rgba(0, 0, 0, 0.4))',
              }}
            >
              <CloseIcon fontSize="large" />
            </IconButton>
            <Button
              aria-label="location submit button"
              variant="contained"
              color={isPositionSelected ? 'success' : 'secondary'}
              onClick={successCallback}
              sx={{
                position: 'absolute',
                bottom: 0,
                left: 0,
                width: '100%',
              }}
            >
              {isPositionSelected ? "I'm here!" : 'Place the marker'}
            </Button>
          </>
        ) : null}
      </Box>
    );
  }
);
