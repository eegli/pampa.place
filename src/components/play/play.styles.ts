import { Box, BoxProps, styled } from '@mui/material';

interface StyledMapOverlayProps extends BoxProps {
  pos: 'map' | 'submit';
}
export const StyledMapOverlay = styled(Box)<StyledMapOverlayProps>(
  ({ pos }) => ({
    zIndex: 10,
    position: 'fixed',
    width: '4rem',
    bottom: pos === 'map' ? 40 : 130,
    right: pos === 'map' ? 70 : 90,
    '&:hover': {
      cursor: 'pointer',
    },
  })
);
