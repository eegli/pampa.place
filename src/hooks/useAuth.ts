import { useCallback, useState } from 'react';
import { AuthReq, AuthRes } from '../pages/api/auth';

export const useAuth = () => {
  console.count('render auth hook');

  const [error, setError] = useState<boolean>(false);

  const authCallback = useCallback(async (authParams: AuthReq) => {
    // First, check if we have a local maps key
    if (process.env.MAPS_LOCAL_API_KEY) {
      return process.env.MAPS_LOCAL_API_KEY;
    }
    try {
      const res: AuthRes = await fetch(
        'api/auth?' + new URLSearchParams(authParams)
      ).then(res => res.json());
      setError(false);
      return res.key;
    } catch (e) {
      setError(true);
    }
  }, []);

  return [authCallback, error] as const;
};
