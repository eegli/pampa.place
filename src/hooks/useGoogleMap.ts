import {RefObject, useEffect} from 'react';
import {LatLngLiteral} from '../config/types';
import {MapService} from '../services/google';

type CenterProps = {
  center: google.maps.LatLng | google.maps.LatLngLiteral;
  zoom?: number;
};

type BoundsProps = {
  SW: LatLngLiteral;
  NE: LatLngLiteral;
};

type Props = {ref: RefObject<HTMLDivElement>; opts?: google.maps.MapOptions} & (
  | CenterProps
  | BoundsProps
);

// TODO make this work
export function useGoogleMap(props: Props) {
  const {ref, ...rest} = props;

  useEffect(() => {
    if (ref.current) {
      console.info('%cMAP MOUNT', 'color: yellow');
      const unmount = MapService.mount(ref.current);

      if ('center' in rest) {
        MapService.map.setCenter(rest.center);
        MapService.map.setZoom(rest.zoom || 3);
      } else {
        /* Order in constructor is important! SW, NE  */
        const bounds = new google.maps.LatLngBounds(
          new google.maps.LatLng(rest.SW),
          new google.maps.LatLng(rest.NE)
        );
        google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
          MapService.map.fitBounds(bounds, 0);
        });
      }

      return () => {
        unmount();
        console.info('%cMAP UNMOUNT', 'color: yellow');
      };
    }
  }, [ref, rest]);
}
