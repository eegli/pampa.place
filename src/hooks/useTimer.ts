import {useEffect, useState} from 'react';

export function useTimer(startTime: number) {
  const [seconds, setSeconds] = useState<number>(startTime);

  useEffect(() => {
    if (seconds > 0) {
      const timer = setTimeout(() => {
        setSeconds(seconds - 1);
      }, 1000);
      return () => clearTimeout(timer);
    }
  }, [seconds]);

  return [seconds] as const;
}
