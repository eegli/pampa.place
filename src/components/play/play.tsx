import {GoogleMap} from '@/components/google/map';
import {GoogleStreetView} from '@/components/google/street-view';
import {MAPS} from '@/config/maps';
import {setPlayerScore} from '@/redux/game';
import {getActivePlayer} from '@/redux/game/selectors';
import {useAppDispatch, useAppSelector} from '@/redux/hooks';
import {Box, Button} from '@mui/material';
import {BoxProps} from '@mui/material/Box';
import {styled} from '@mui/material/styles';
import {useState} from 'react';
import {PlayHeader} from './play-header';

interface StyledMapOverlayProps extends BoxProps {
  pos: 'map' | 'submit';
}
const StyledMapOverlay = styled(Box)<StyledMapOverlayProps>(({pos}) => ({
  zIndex: 10,
  position: 'fixed',
  width: '4rem',
  bottom: pos === 'map' ? 40 : 130,
  right: pos === 'map' ? 70 : 90,
  '&:hover': {
    cursor: 'pointer',
  },
}));

export const Play = () => {
  const dispatch = useAppDispatch();
  const [showMap, setShowMap] = useState<boolean>(false);

  const activePlayer = useAppSelector(getActivePlayer);
  const initialPos = useAppSelector(({position}) => position.initialPosition);
  const selectedPos = useAppSelector(({position}) => position.selectedPosition);
  const activeMapId = useAppSelector(({game}) => game.mapId);
  const map = MAPS.get(activeMapId);

  function submitScore() {
    dispatch(
      setPlayerScore({
        initial: initialPos,
        selected: selectedPos,
      })
    );
  }

  if (!map || !initialPos) {
    return null;
  }

  return (
    <Box
      id="play-mode"
      sx={{
        display: 'flex',
        flexFlow: 'column',
        alignItems: 'center',
        height: '100%',
        width: '100%',
        overflow: 'hidden',
      }}
    >
      <PlayHeader player={activePlayer} timerCallback={submitScore} />
      <Box position="relative" height="100%" width="100%">
        <StyledMapOverlay pos="map" onClick={() => setShowMap(!showMap)}>
          {/* eslint-disable-next-line  @next/next/no-img-element */}
          <img aria-label="map-toggle" src="/map.svg" alt="map-icon" />
        </StyledMapOverlay>

        <StyledMapOverlay pos="submit" onClick={() => setShowMap(!showMap)}>
          <Button
            size={'large'}
            variant="contained"
            color="secondary"
            onClick={submitScore}
            disabled={!selectedPos}
          >
            Submit
          </Button>
        </StyledMapOverlay>
        <GoogleStreetView display={showMap ? 'none' : 'block'} />
        <GoogleMap map={map} mode="play" />
      </Box>
    </Box>
  );
};
