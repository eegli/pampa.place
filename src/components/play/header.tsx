import {useAppSelector} from '@/redux/hooks';
import {formatDur} from '@/utils/misc';
import {Box, Divider, Stack, Typography} from '@mui/material';
import {useRouter} from 'next/router';
import {useEffect} from 'react';
import {useTimer} from '../../hooks/useTimer';

type PlayHeaderProps = {
  player: string;
  timerCallback: () => void;
};

export const PlayHeader = ({player, timerCallback}: PlayHeaderProps) => {
  const router = useRouter();
  const timeLimit = useAppSelector(s => s.game.timeLimit);
  const [timeRemaining] = useTimer(timeLimit);

  const isUnlimitedTimeMode = timeLimit < 0;

  const label = isUnlimitedTimeMode ? (
    <span>&#8734;</span>
  ) : (
    <span>{formatDur(timeRemaining)}</span>
  );

  useEffect(() => {
    if (!isUnlimitedTimeMode && !timeRemaining) {
      timerCallback();
    }
  }, [timeRemaining, isUnlimitedTimeMode, timerCallback]);

  function handleClick() {
    router.push('/');
  }

  return (
    <Box px={2} my={2} width="100%" display="flex">
      <Stack
        flexGrow={1}
        display="flex"
        justifyContent="center"
        direction="row"
        divider={<Divider orientation="vertical" flexItem />}
        spacing={2}
      >
        <Box textAlign="right">
          <Typography
            variant="caption"
            color={'text.secondary'}
            sx={{
              textTransform: 'uppercase',
            }}
          >
            Your turn
          </Typography>
          <Typography variant="h5">{player}</Typography>
        </Box>
        <Box>
          <Typography
            variant="caption"
            color={'text.secondary'}
            sx={{
              textTransform: 'uppercase',
            }}
          >
            Time remaining
          </Typography>
          <Typography variant="h5" color={'secondary.main'}>
            {label}
          </Typography>
        </Box>
      </Stack>
    </Box>
  );
};
