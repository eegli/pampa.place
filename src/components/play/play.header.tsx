import { useAppSelector } from '@/redux/hooks';
import { formatDuration } from '@/utils/misc';
import HomeIcon from '@mui/icons-material/Home';
import { Box, Divider, IconButton, Stack, Typography } from '@mui/material';
import { useRouter } from 'next/router';
import { useEffect } from 'react';
import { useTimer } from '../../hooks/useTimer';

type PlayHeaderProps = {
  player: string;
  timerCallback: () => void;
};

const getLabel = (unlimited: boolean, time: number) => {
  return unlimited ? <span>&#8734;</span> : <span>{formatDuration(time)}</span>;
};

const PlayHeader = ({ player, timerCallback }: PlayHeaderProps) => {
  const router = useRouter();
  const timeLimit = useAppSelector(s => s.game.timeLimit);
  const [timeRemaining] = useTimer(timeLimit);

  const isUnlimitedTimeMode = timeLimit < 0;

  const label = getLabel(isUnlimitedTimeMode, timeRemaining);

  useEffect(() => {
    if (!timeRemaining && !isUnlimitedTimeMode) {
      timerCallback();
    }
  }, [timeRemaining, isUnlimitedTimeMode, timerCallback]);

  function handleClick() {
    router.push('/');
  }

  return (
    <Box px={2} my={2} width="100%" display="flex">
      <Stack
        flexGrow={1}
        display="flex"
        justifyContent="center"
        direction="row"
        divider={<Divider orientation="vertical" flexItem />}
        spacing={2}
      >
        <Box textAlign="right">
          <Typography
            variant="caption"
            color={'text.secondary'}
            sx={{
              textTransform: 'uppercase',
            }}
          >
            Your turn
          </Typography>
          <Typography variant="h5">{player}</Typography>
        </Box>
        <Box>
          <Typography
            variant="caption"
            color={'text.secondary'}
            sx={{
              textTransform: 'uppercase',
            }}
          >
            Time remaining
          </Typography>
          <Typography variant="h5" color={'secondary.main'}>
            {label}
          </Typography>
        </Box>
      </Stack>

      <IconButton sx={{ mx: 2 }} size="large" edge="end" onClick={handleClick}>
        <HomeIcon />
      </IconButton>
    </Box>
  );
};
export default PlayHeader;
