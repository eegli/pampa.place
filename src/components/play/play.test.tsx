import {
  createMockState,
  createMockStore,
  fireEvent,
  render,
  screen,
} from '@/tests/utils';
import * as timer from '../../hooks/useTimer';
import {Play} from './play';

/* WORK IN PROGRESS */

describe('Play', () => {
  function getSubmitButton() {
    return screen.getByRole('button', {name: /submit/i});
  }
  it('has disabled submit button state', () => {
    const state = createMockState({
      position: {
        selectedPosition: null,
      },
    });
    const store = createMockStore(state);
    render(<Play />, store);
    const button = getSubmitButton();
    expect(button).toBeDisabled();
    expect(button).toBeInTheDocument();
  });
  it('has enabled button state', () => {
    const state = createMockState({
      position: {
        selectedPosition: {
          lat: 1,
          lng: 1,
        },
      },
    });
    const store = createMockStore(state);
    render(<Play />, store);
    const button = getSubmitButton();
    expect(button).not.toBeDisabled();
    expect(button).toBeInTheDocument();
  });
  it('dispatches no score when time runs out', () => {
    const useTimerSpy = jest.spyOn(timer, 'useTimer').mockImplementation(() => {
      return [0] as const;
    });
    const state = createMockState({
      position: {
        selectedPosition: null,
        initialPosition: {
          lat: 2,
          lng: 2,
        },
      },
      game: {
        players: ['player 1'],
        scores: [[]],
      },
    });
    const store = createMockStore(state);
    render(<Play />, store);
    expect(store.getState().game.scores).toMatchSnapshot('no score');
    useTimerSpy.mockRestore();
  });
  it('dispatches score when position is set', () => {
    const state = createMockState({
      position: {
        selectedPosition: {
          lat: 1,
          lng: 1,
        },
        initialPosition: {
          lat: 2,
          lng: 2,
        },
      },
      game: {
        players: ['player 1'],
        scores: [[]],
      },
    });
    const store = createMockStore(state);
    render(<Play />, store);
    fireEvent.click(getSubmitButton());
    expect(store.getState().game.scores).toMatchSnapshot('score');
  });
});
