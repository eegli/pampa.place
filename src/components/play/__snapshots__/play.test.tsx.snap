// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play dispatches no score when time runs out: no score 1`] = `
Array [
  Array [
    Object {
      "dist": -1,
      "name": "player 1",
      "score": 0,
      "selected": null,
    },
  ],
]
`;

exports[`Play dispatches score when position is set: score 1`] = `
Array [
  Array [
    Object {
      "dist": 157225.64920708904,
      "name": "player 1",
      "score": 65,
      "selected": Object {
        "lat": 1,
        "lng": 1,
      },
    },
  ],
]
`;
